uint16[] ports
bool[] states
